examples/basic_spec.rs

```diff
@@ -19,26 +19,27 @@ fn basic_spec() -> ParserSpec<String> {
     
     let mut spec = ParserSpec::new();
     
-    spec.add_null_associations(vec!["ident", "number"], PrecedenceLevel::Root, |_, token, _| {Ok(Node::Simple(token))});
-    spec.add_left_associations(vec!["+", "-"],  PrecedenceLevel::First, recurse_call);
-    spec.add_left_associations(vec!["*", "/", "%"], PrecedenceLevel::Second, recurse_call);
-    spec.add_null_assoc("(", PrecedenceLevel::First, parens_call);
-    spec
+    spec.add_null_associations(vec!["ident", "number"], PrecedenceLevel::Root, |_, token, _| {Ok(Node::Simple(token))}) ?;
+    spec.add_left_associations(vec!["+", "-"],  PrecedenceLevel::First, recurse_call)?;
+    spec.add_left_associations(vec!["*", "/", "%"], PrecedenceLevel::Second, recurse_call)?;
+    spec.add_null_assoc("(", PrecedenceLevel::First, parens_call)?;
+    Ok(spec)
 }
 
 
-pub struct BasicParser {
+pub struct BasicParser<L: Lexer<String>> {
     null_map: HashMap<String, NullInfo<String>>, 
     left_map: HashMap<String, LeftInfo<String>>,
-    lexer: Box<Lexer<String>>, 
+    lexer: L
```

Comment:
It's considered better to feed input to parsers from the outside rather than them producing it themselves, because this allows them to be oblivious to async concerns. I don't know what changes would have to be made to the parsing algorithm to support this however.